notes/inheritance-1.md

```diff
@@ -0,0 +1,351 @@
+
+Example:
+```C++
+Bun b;
+```
+- Variable `b`'s static type is `Bun`, its dynamic type is also `Bun`
+- Another way to put it: a variable of type `Bun` is always a `Bun` object
+
+<mark>However, for a pointer or reference, the dynamic type depends on the type of argument to which the pointer/reference is bound</mark>
+
+Example:
+```C++
+void printRank(Animal &ani){
+    // ani's static type is Animal&
+    std::cout << ani.Rank() << std::endl;
+}
+
+int main(){
+    Animal a;
+    Bun b;
+    KingBun kb;
+    printRank(a); // ani's dynamic type is Animal
+    printRank(b); // ani's dynamic type is Bun
+    printRank(kb); // ani's dynamic type is KingBun
+}
+
+```
+The above example is important!
+- We said earlier that `printRank`'s parameter `ani` has static type `Animal&`
+- However, its dynamic type can be `Animal`, `Bun`, or `KingBun`
+
+---
+## Base to derived conversion?
+
+You might be wondering: in C++, is it safe to implicitly convert from a base class object to a derived class object?
+
+Generally no. Think back to our section on [subobjects](#derived-to-base-conversions). Most of the time, there is no guarantee that a base class object contains the members defined by the derived class.
+- Conceptual example: Say you needed the help of [Korosensei](https://ansatsukyoshitsu.fandom.com/wiki/Korosensei), an excellent teacher who can also fly at top speeds of Mach 20. Since I can't guarantee that any random teacher can also fly, I'm not allowed to perform a conversion
+
+Thus, there is no implicit conversion from base to derived
+```C++
+Animal a;
+Bun* bunPtr = &a;  // compile error, cannot convert base to derived
+Bun& bunRef = a; // compile error
+
+```
+
+
+---
+## Polymorphism
+
+The key concept behind this entire chapter is polymorphism
+
+Through polymorphism, a single entity (e.g., a variable, function, or object), can have multiple forms. Most importantly, this allows us to treat objects, which may be of different types, in the same exact way
+- A whole host of benefits arises from using polymorphism: less redundancy, more understandable/modifiable code, etc.
+
+We used this concept to great effect in the previous example. Instead of writing the same function for each object of base type `Animal`, we wrote code that allows us to treat all objects of type `Animal` or one of its derived classes in the same exact way!
+
+---
+## Types of polymorphism
+
+There are 2 main types of polymorphism: static polymorphism and dynamic polymorphism
+
+You've probably seen a lot of **static polymorphism**, also called **compile-time polymorphism**. This is just regular function overloading!
+- When multiple functions have the same name but different signatures, the compiler will pick which one to call based on the arguments passed
+- This happens at compile-time
+- This is what we abused when we wrote the redundant version of `printRank`
```

Comment:
Can you clarify which example here? It sort of took me a while to realize that it was not the one I just saw (Or actually, it might just be me).

Reply:
good catch, done